### PR TITLE
perf: reduce String allocation and use Unsafe to optimize performance

### PR DESCRIPTION
**Pull request by LinShunKang**
Motivation:

Reduce memory allocation and optimize performance for `QueryStringDecoder`.

Modification:

* Replace `new String()` with `Utf8Utils.decodeUtf8` when charset is UTF-8.
* Use `Unsafe` to optimize performance.
* Use `char[]` instead of `StringBuilder` to optimize performance

Result:

Before:
```
Benchmark                                         Mode  Cnt  Score   Error   Units
QueryStringDecoderBenchmark.mixedDecoding        thrpt    6  1.380 ± 0.140  ops/us
QueryStringDecoderBenchmark.noDecoding           thrpt    6  6.670 ± 0.236  ops/us
QueryStringDecoderBenchmark.nonStandardDecoding  thrpt    6  1.825 ± 1.031  ops/us
QueryStringDecoderBenchmark.onlyDecoding         thrpt    6  1.549 ± 0.146  ops/us
```

After:
```
Benchmark                                         Mode  Cnt  Score   Error   Units
QueryStringDecoderBenchmark.mixedDecoding        thrpt    6  1.728 ± 0.313  ops/us
QueryStringDecoderBenchmark.noDecoding           thrpt    6  6.567 ± 0.997  ops/us
QueryStringDecoderBenchmark.nonStandardDecoding  thrpt    6  1.892 ± 0.405  ops/us
QueryStringDecoderBenchmark.onlyDecoding         thrpt    6  2.040 ± 0.076  ops/us
```

### CODE REVIEW

**Comment by LinShunKang** [comment]
@normanmaurer PTAL

**Comment by franz1981** [comment]
It looks to me that we have many changes here  and I am not sure they worth it, given that the original benchmark is already quite intensive GC-wise and it's focused on always decoding the same strings. 
In addition, beware using unsafe to save bound checks because depending on the cases you would loose any auto-vectorization chance (if there is some).
I would proceed like this:
1) cleanup/modify a bit the bench to be sure that it measures what you're expecting to stress
2) we really need that many new utf8utils methods or there is something we already have on Netty to be improved/merged/reused?

These are my 2 cents on this NS thanks for the PR and the hard work on it :)

File: codec-http/src/main/java/io/netty/handler/codec/http/Utf8Utils.java [hunk 1]
```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2012 The Netty Project
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/Utf8Utils.java]: 2020

File: codec-http/src/main/java/io/netty/handler/codec/http/Utf8Utils.java [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+import static java.lang.Character.MIN_HIGH_SURROGATE;
+import static java.lang.Character.MIN_LOW_SURROGATE;
+import static java.lang.Character.MIN_SUPPLEMENTARY_CODE_POINT;
+
+/**
+ * See original <a
+ * href="https://github.com/protocolbuffers/protobuf/blob/master/java/core/src/main/java/com/google/protobuf/Utf8.java">
+ * Utf8.java</a>
+ */
+public final class Utf8Utils {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/Utf8Utils.java]: Seems like this is based on a protobuf class. We need to retain the copyright header etc and also add it to the NOTICE file.

**Comment by normanmaurer** [comment]
I agree with @franz1981 here... Honestly I am not sure it is worth to maintain all the code for this

**Comment by LinShunKang** [comment]
> It looks to me that we have many changes here and I am not sure they worth it, given that the original benchmark is already quite intensive GC-wise and it's focused on always decoding the same strings.
> In addition, beware using unsafe to save bound checks because depending on the cases you would loose any auto-vectorization chance (if there is some).
> I would proceed like this:
> 
> 1. cleanup/modify a bit the bench to be sure that it measures what you're expecting to stress
> 2. we really need that many new utf8utils methods or there is something we already have on Netty to be improved/merged/reused?
> 
> These are my 2 cents on this: thanks for the PR and the hard work on it :)

I tried to find UTF-8 encode methods in Netty, but I didn't find anything, so I modify the UTF-8 encode methods in protobuf.

**Comment by LinShunKang** [comment]
Add new benchmark `longMixedDecoding`, result:
Before:
```
Benchmark                                         Mode  Cnt     Score     Error   Units
QueryStringDecoderBenchmark.longMixedDecoding    thrpt    6   270.983 ±  38.867  ops/ms
QueryStringDecoderBenchmark.mixedDecoding        thrpt    6  1460.959 ± 227.085  ops/ms
QueryStringDecoderBenchmark.noDecoding           thrpt    6  7310.070 ±  62.510  ops/ms
QueryStringDecoderBenchmark.nonStandardDecoding  thrpt    6  1588.976 ±  42.605  ops/ms
QueryStringDecoderBenchmark.onlyDecoding         thrpt    6  1802.212 ± 508.778  ops/ms
```

After:
```
Benchmark                                         Mode  Cnt     Score     Error   Units
QueryStringDecoderBenchmark.longMixedDecoding    thrpt    6   289.256 ±  42.580  ops/ms
QueryStringDecoderBenchmark.mixedDecoding        thrpt    6  1638.161 ± 136.633  ops/ms
QueryStringDecoderBenchmark.noDecoding           thrpt    6  7159.545 ± 393.757  ops/ms
QueryStringDecoderBenchmark.nonStandardDecoding  thrpt    6  1644.291 ±  63.117  ops/ms
QueryStringDecoderBenchmark.onlyDecoding         thrpt    6  2161.531 ±  76.752  ops/ms
```

**Comment by LinShunKang** [comment]
Use char[] instead of StringBuilder to optimize performance, result from https://github.com/netty/netty/pull/10092/commits/f724ee644647f4a1d4649dbc2d454964cf02483d

Before:
```
Benchmark                                         Mode  Cnt     Score     Error   Units
QueryStringDecoderBenchmark.longMixedDecoding    thrpt    6   270.983 ±  38.867  ops/ms
QueryStringDecoderBenchmark.mixedDecoding        thrpt    6  1460.959 ± 227.085  ops/ms
QueryStringDecoderBenchmark.noDecoding           thrpt    6  7310.070 ±  62.510  ops/ms
QueryStringDecoderBenchmark.nonStandardDecoding  thrpt    6  1588.976 ±  42.605  ops/ms
QueryStringDecoderBenchmark.onlyDecoding         thrpt    6  1802.212 ± 508.778  ops/ms
```

After
```
Benchmark                                         Mode  Cnt     Score      Error   Units
QueryStringDecoderBenchmark.longMixedDecoding    thrpt    6   370.368 ±    3.704  ops/ms
QueryStringDecoderBenchmark.mixedDecoding        thrpt    6  2009.134 ±   23.247  ops/ms
QueryStringDecoderBenchmark.noDecoding           thrpt    6  6946.912 ± 1306.370  ops/ms
QueryStringDecoderBenchmark.nonStandardDecoding  thrpt    6  2204.528 ±  296.804  ops/ms
QueryStringDecoderBenchmark.onlyDecoding         thrpt    6  2690.175 ±   51.252  ops/ms
```

**Comment by LinShunKang** [comment]
Use `FastThreadLocal` to reduce `byte[]` and `char[]` allocation, result from https://github.com/netty/netty/pull/10092/commits/b7a374587c32aa3763dd4d4c44b204e1de931938:

Original:
```
Benchmark                                         Mode  Cnt     Score     Error   Units
QueryStringDecoderBenchmark.longMixedDecoding    thrpt    6   270.983 ±  38.867  ops/ms
QueryStringDecoderBenchmark.mixedDecoding        thrpt    6  1460.959 ± 227.085  ops/ms
QueryStringDecoderBenchmark.noDecoding           thrpt    6  7310.070 ±  62.510  ops/ms
QueryStringDecoderBenchmark.nonStandardDecoding  thrpt    6  1588.976 ±  42.605  ops/ms
QueryStringDecoderBenchmark.onlyDecoding         thrpt    6  1802.212 ± 508.778  ops/ms
```

After:
```
Benchmark                                         Mode  Cnt     Score     Error   Units
QueryStringDecoderBenchmark.longMixedDecoding    thrpt    6   390.395 ±  73.598  ops/ms
QueryStringDecoderBenchmark.mixedDecoding        thrpt    6  1554.062 ± 186.379  ops/ms
QueryStringDecoderBenchmark.noDecoding           thrpt    6  6183.945 ± 315.638  ops/ms
QueryStringDecoderBenchmark.nonStandardDecoding  thrpt    6  2256.158 ±  83.484  ops/ms
QueryStringDecoderBenchmark.onlyDecoding         thrpt    6  2602.271 ± 914.358  ops/ms
```

**Comment by LinShunKang** [comment]
@franz1981 I optimize `QueryStringDecoder` because `QueryStringDecoder` allocate too much memory:

Before optimize：
````
cn.shade.gateway.core.codec.QueryStringDecoder.decodeComponent: 7,324,670,728 (8.56%) bytes in 178,510,676 (12.38%) objects (avg size 41 bytes)
        char[]: 4,675,845,776 (5.46%) bytes in 86,376,625 (5.99%) objects (avg size 54 bytes)
        java.lang.String: 2,003,941,416 (2.34%) bytes in 83,497,559 (5.79%) objects (avg size 24 bytes)
        byte[]: 368,531,072 (0.43%) bytes in 2,879,149 (0.19%) objects (avg size 128 bytes)
        java.nio.HeapCharBuffer: 138,176,304 (0.16%) bytes in 2,878,673 (0.19%) objects (avg size 48 bytes)
        java.nio.HeapByteBuffer: 138,176,160 (0.16%) bytes in 2,878,670 (0.19%) objects (avg size 48 bytes)
````

After optimize：
````
cn.shade.gateway.core.codec.QueryStringDecoder.decodeComponent: 2,082,722,280 (2.36%) bytes in 56,660,130 (3.99%) objects (avg size 37 bytes)
        char[]: 1,402,956,216 (1.59%) bytes in 28,336,544 (1.99%) objects (avg size 50 bytes)
        java.lang.String: 679,766,064 (0.77%) bytes in 28,323,586 (1.99%) objects (avg size 24 bytes)
````

The original `QueryStringDecoder` allocated 7,324MB memory, but the optimized `QueryStringDecoder` only allocated 2,082MB memory.

PS: 
* `cn.shade.gateway.core.codec.QueryStringDecoder` is a base on Netty (before 4.1.45) `QueryStringDecoder`
* These information collected by aprof

**Comment by franz1981** [comment]
Thanks for the explanation.
Re
> because QueryStringDecoder allocate too much memory

Yes, but still, just once per QueryStringDecoder: the benchmark is promoting a non realistic usage of it, just for the sake of measuring the decoding cost (and I don't agree on it TBH, I would have written the benchmark differently).
If your use case shows that QueryStringDecoder::parameters is a bottleneck for your use case I suggest to focus on reduce the amount of garbage produced without the need of that many changes anyway (maybe trying to understand why many intermediate allocations are not correctly escaped?).

**Comment by LinShunKang** [comment]
These information collected by aprof, it was use wrk rather than JMH to benchmark, and all http requests were realistic.

And 
> maybe trying to understand why many intermediate allocations are not correctly escaped?

Sorry, I don't understand these words.

**Comment by franz1981** [comment]
> maybe trying to understand why many intermediate allocations are not correctly escaped?

The original version was using some intermediate classes that I would expect them to not be allocated if escape analysis would have worked correctly: why it isn't happening?

In addition: if QueryStringDecoder::parameters represent your bottleneck probably the way it encodes the parameters is too expensive (map of list of strings). I would rather try to fix the root cause of the issue instead of proposing many changes just to alleviate one synthom.

**Comment by LinShunKang** [comment]
About "Why Escape Analysis is not happening", it is a good question, but I do not have any ideas now.

Yes, I agree with you about encodes parameters is too expensive too, and this is the root cause of the issue, but this work is too expensive to implement(It need change the agreement between App, ApiGateway and RealServer).

**Comment by normanmaurer** [comment]
I am with @franz1981 still... I am not convinced we want to maintain the extra code for the win here.

**Comment by LinShunKang** [comment]
OK : (

**Comment by franz1981** [comment]
Don't worry and thanks for the contribution: I see lot of potential to do very good stuff here don't get down about this!

**Pull request closed without merging**